core/amber/src/main/scala/edu/uci/ics/texera/workflow/common/workflow/WorkflowCompiler.scala

```diff
@@ -15,21 +15,105 @@ import edu.uci.ics.texera.web.workflowruntimestate.WorkflowAggregatedState.FAILE
+    */
+  def compileToLogicalPlan(
+      logicalPlanPojo: LogicalPlanPojo
+  ): (LogicalPlan, Map[OperatorIdentity, WorkflowFatalError]) = {
+    val errorList = new ArrayBuffer[(OperatorIdentity, Throwable)]()
+    val opIdToError = mutable.Map[OperatorIdentity, WorkflowFatalError]()
+
+    var logicalPlan: LogicalPlan = LogicalPlan(logicalPlanPojo)
+    logicalPlan = SinkInjectionTransformer.transform(
+      logicalPlanPojo.opsToViewResult,
+      logicalPlan
+    )
+
+    logicalPlan.propagateWorkflowSchema(context, Some(errorList))
+    // report compilation errors
+    if (errorList.nonEmpty) {
+      errorList.foreach {
+        case (opId, err) =>
+          logger.error("error occurred in logical plan compilation", err)
+          opIdToError += (opId -> WorkflowFatalError(
+            COMPILATION_ERROR,
+            Timestamp(Instant.now),
+            err.toString,
+            getStackTraceWithAllCauses(err),
+            opId.id
+          ))
+      }
+    }
+    (logicalPlan, opIdToError.toMap)
+  }
+
+  /**
+    * Compile a workflow to physical plan, along with the schema propagation result and error(if any)
+    *
+    * @param logicalPlanPojo the pojo parsed from workflow str provided by user
+    * @return WorkflowCompilationResult, containing the physical plan, input schemas per op and error per op
+    */
+  def compileToPhysicalPlan(
+                             logicalPlanPojo: LogicalPlanPojo
+                           ): WorkflowCompilationResult = {
+    val (logicalPlan, opIdToError) = compileToLogicalPlan(logicalPlanPojo)
+    if (opIdToError.nonEmpty) {
+      // encounter errors during compile pojo to logical plan,
+      //   so directly return empty physical plan, schema map and non-empty error map
+      return WorkflowCompilationResult(
+        physicalPlan = PhysicalPlan.empty,
+        operatorIdToInputSchemas = Map.empty,
+        operatorIdToError = opIdToError
+      )
+    }
+    // from logical plan to physical plan
+    val physicalPlan = PhysicalPlan(context, logicalPlan)
+
+    // Extract physical input schemas, excluding internal ports
+    val physicalInputSchemas = physicalPlan.operators.map { physicalOp =>
+      physicalOp.id -> physicalOp.inputPorts.values
+        .filterNot(_._1.id.internal)
+        .map {
+          case (port, _, schema) => port.id -> schema.toOption
+        }
+    }
+
+    // Group the physical input schemas by their logical operator ID and consolidate the schemas
+    val opIdToInputSchemas = physicalInputSchemas
+      .groupBy(_._1.logicalOpId)
+      .view
+      .mapValues(_.flatMap(_._2).toList.sortBy(_._1.id).map(_._2))
+      .toMap
+
+    WorkflowCompilationResult(physicalPlan, opIdToInputSchemas, Map.empty)
+  }
+
   def compileLogicalPlan(
       logicalPlanPojo: LogicalPlanPojo,
       executionStateStore: ExecutionStateStore
   ): LogicalPlan = {
-
+    // TODO: remove this function after separating compiler as a standalone service
```

Comment:
use function block comment, instead of line comment of the first line of function.

Reply:
Done